### PR TITLE
feat: Implements `merge_mode` region options

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?
This PR implements a new region option `update_mode` to control how to update existing rows by key. There are two update modes available:
- `LastRow`: overwrites all fields by the values in the latest row
- `LastNotNull`: only overwrites the null value by the latest value for each field

Memtables have individual dedup implementations so they can't reuse the `DedupReader`. To avoid repeating the dedup logic for each memtable, this PR implements a new iterator `LastNotNullIter` as a wrapper for the memtable iter. `LastNotNullIter` splits batches and invokes the `LastNotNull` dedup strategy to dedup them. The `LastNotNullIter` only takes effect when `update_mode` is `LastNotNull`.

After this PR, I will expose the option to table options and add some sqlness tests.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced `MergeMode` options for handling duplicate row merging in various components.
  - Added new iterator `LastNotNullIter` for deduplication based on the `LastNotNull` strategy.

- **Enhancements**
  - Updated time series operations and compaction processes to utilize the new `MergeMode` options.
  - Enhanced `TimeSeriesMemtable`, `PartitionTreeConfig`, and `ScanInput` to include `merge_mode` configuration.

- **Tests**
  - Added comprehensive tests for new `MergeMode` options and append mode functionality.

### CODE REVIEW

File: src/mito2/src/memtable/partition_tree.rs [hunk 1]
```diff
@@ -98,6 +102,7 @@ impl Default for PartitionTreeConfig {
             data_freeze_threshold: 131072,
             dedup: true,
             fork_dictionary_bytes,
+            update_mode: UpdateMode::LastRow,
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/memtable/partition_tree.rs:105]: `PartitionTree::read()` will use this.

File: src/mito2/src/engine/update_mode_test.rs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    let batches = RecordBatches::try_collect(stream).await.unwrap();
+    let expected = "\
++-------+---------+---------+---------------------+
+| tag_0 | field_0 | field_1 | ts                  |
++-------+---------+---------+---------------------+
+| a     | 11.0    |         | 1970-01-01T00:00:01 |
+| a     | 22.0    | 222.0   | 1970-01-01T00:00:02 |
+| a     | 13.0    | 3.0     | 1970-01-01T00:00:03 |
++-------+---------+---------+---------------------+";
+    assert_eq!(expected, batches.pretty_print().unwrap());
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/engine/update_mode_test.rs:85]: Some edge cases are tested by unit tests for `LastRow` and `LastNotNull`.

File: src/mito2/src/memtable/partition_tree/tree.rs [hunk 3]
```diff
@@ -83,9 +85,13 @@ impl PartitionTree {
                 .collect(),
         };
         let is_partitioned = Partition::has_multi_partitions(&metadata);
+        let mut config = config.clone();
+        if config.update_mode == UpdateMode::LastNotNull {
+            config.dedup = false;
+        }
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/memtable/partition_tree/tree.rs:91]: This is a little verbose.

**Comment by killme2008** [comment]
I am not convinced that `UpdateMode` is an appropriate term. It is utilized during scanning to combine duplicate rows, but referred to as `UpdateMode`. Personally, I would like `MergeMode` or `ScanMergeMode` like that.

File: src/mito2/src/read/dedup.rs [hunk 4]
```diff
@@ -982,4 +1076,92 @@ mod tests {
...
+        let actual: Vec<_> = iter.map(|batch| batch.unwrap()).collect();
+        let expect = [
+            new_batch_multi_fields(b"k1", &[1], &[13], &[OpType::Put], &[(Some(1), None)]),
+            new_batch_multi_fields(b"k1", &[2], &[13], &[OpType::Put], &[(Some(2), Some(22))]),
+            new_batch_multi_fields(b"k1", &[3], &[13], &[OpType::Delete], &[(None, None)]),
+            new_batch_multi_fields(b"k2", &[1], &[13], &[OpType::Put], &[(Some(1), None)]),
+            new_batch_multi_fields(b"k2", &[2], &[13], &[OpType::Put], &[(Some(2), Some(22))]),
+        ];
+        assert_eq!(&expect, &actual[..]);
+    }
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/mito2/src/read/dedup.rs:1166]: @CookiePieWw please try it

**Comment by fengjiachun** [comment]
> I am not convinced that `UpdateMode` is an appropriate term. It is utilized during scanning to combine duplicate rows, but referred to as `UpdateMode`. Personally, I would like `MergeMode` or `ScanMergeMode` like that.

`UpdateMode` seems to be viewed from the perspective of writing. cc @evenyag

File: src/mito2/src/read/dedup.rs [hunk 5]
```diff
@@ -468,6 +468,100 @@ impl DedupStrategy for LastNotNull {
...
+    /// If no duplicate rows, returns None.
+    fn find_split_index(batch: &Batch) -> Option<usize> {
+        if batch.num_rows() < 2 {
+            return None;
+        }
+
+        // Safety: The batch is not empty.
+        let timestamps = batch.timestamps_native().unwrap();
+        timestamps.windows(2).position(|t| t[0] == t[1])
+    }
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/read/dedup.rs:508]: This is unreasonable.

File: src/mito2/src/read/dedup.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/mito2/src/read/dedup.rs:1166]: @coderabbitai generate additional test cases

**Comment by evenyag** [comment]
> > I am not convinced that `UpdateMode` is an appropriate term. It is utilized during scanning to combine duplicate rows, but referred to as `UpdateMode`. Personally, I would like `MergeMode` or `ScanMergeMode` like that.
> 
> `UpdateMode` seems to be viewed from the perspective of writing. cc @evenyag

Good idea. `MergeMode` may be proper.

File: src/mito2/src/read/dedup.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (CookiePieWw) [reply src/mito2/src/read/dedup.rs]: > @CookiePieWw please try it

@killme2008 I'm not quite in the loop here. Is there anything I can help with? (or it is a misspelling with my name and the code rabbit :)

**Comment by killme2008** [comment]
@evenyag Don't forget to change the PR title and content.

File: src/mito2/src/compaction.rs [hunk 6]
```diff
@@ -453,31 +454,39 @@ pub struct SerializedCompactionOutput {
...
+        let mut scan_input = ScanInput::new(self.sst_layer, ProjectionMapper::all(&self.metadata)?)
+            .with_files(self.inputs.to_vec())
+            .with_append_mode(self.append_mode)
+            .with_cache(self.cache)
+            .with_filter_deleted(self.filter_deleted)
+            // We ignore file not found error during compaction.
+            .with_ignore_file_not_found(true)
+            .with_merge_mode(self.merge_mode);
+
+        // This serves as a workaround of https://github.com/GreptimeTeam/greptimedb/issues/3944
```
Comment (killme2008) [inline src/mito2/src/compaction.rs:481]: BTW. How can we fix this workaround? @v0y4g3r

File: src/mito2/src/engine/append_mode_test.rs [hunk 7]
```diff
@@ -113,7 +113,7 @@ async fn test_append_mode_compaction() {
         .await
         .unwrap();
 
-    // Flush 2 SSTs for compaction.
+    // Flush 3 SSTs for compaction.
```
Comment (killme2008) [inline src/mito2/src/engine/append_mode_test.rs:116]: was the comment incorrect or we change this test?

File: src/mito2/src/engine/merge_mode_test.rs [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+| tag_0 | field_0 | field_1 | ts                  |
++-------+---------+---------+---------------------+
+| a     | 11.0    |         | 1970-01-01T00:00:01 |
+| a     | 22.0    | 222.0   | 1970-01-01T00:00:02 |
+| a     | 13.0    | 3.0     | 1970-01-01T00:00:03 |
++-------+---------+---------+---------------------+";
+    assert_eq!(expected, batches.pretty_print().unwrap());
+}
+
+#[tokio::test]
```
Comment (killme2008) [inline src/mito2/src/engine/merge_mode_test.rs:88]: Looks like it doesn't test `last_row` mode.

File: src/mito2/src/read/dedup.rs [hunk 9]
```diff
@@ -462,12 +462,106 @@ impl DedupStrategy for LastNotNull {
...
+            // No current batch. Fetches a new batch from the inner iterator.
+            let Some(iter) = self.iter.as_mut() else {
+                // The iterator is exhausted.
+                return Ok(None);
+            };
+
+            self.current_batch = iter.next().transpose()?;
+            if self.current_batch.is_none() {
+                // The iterator is exhausted.
+                self.iter = None;
```
Comment (killme2008) [inline src/mito2/src/read/dedup.rs:525]: should it return `Ok(None)` when `self.current_batch.is_none()` too? Fast path.

File: src/mito2/src/region/options.rs [hunk 10]
```diff
@@ -54,6 +67,9 @@ pub struct RegionOptions {
     pub index_options: IndexOptions,
     /// Memtable options.
     pub memtable: Option<MemtableOptions>,
+    /// The mode to merge duplicate rows.
+    /// Only takes effect when `append_mode` is `false`.
+    pub merge_mode: MergeMode,
```
Comment (killme2008) [inline src/mito2/src/region/options.rs:72]: I think it's better to validate this constraint when creating or opening regions.

**Review by killme2008 — COMMENTED** [review]
Good job. The final piece to support InfluxDB line protocol.

File: src/mito2/src/engine/append_mode_test.rs (hunk 7 above)
Comment (evenyag) [reply src/mito2/src/engine/append_mode_test.rs:116]: The comment is outdated.

File: src/mito2/src/engine/merge_mode_test.rs (hunk 8 above)
Comment (evenyag) [reply src/mito2/src/engine/merge_mode_test.rs:88]: `LastRow` mode is the default option so existing tests already test it.

**Comment by evenyag** [comment]
@coderabbitai pause

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/mito2/src/compaction.rs (hunk 6 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction.rs:481]: > BTW. How can we fix this workaround? @v0y4g3r

Adding a page filter in current parquet reader to prune pages should be enough. But we need to evaluate the performance in that timestamps in pages is not always monotonic, page filter is fast, but it may not always work. I'll reopen that issue once I have a conclusion.
